Async card metadata

dpsutton opened the pull request:
Resolves #14957

#### Background:
We need query metadata on cards. This is how the FE knows how to display the results, but it is also how the query processor can use nested queries. Without this information we cannot select from another query (we don't know what to select). This makes metadata _extremely_ important to have.

This also causes some problems. When we are saving questions we have to get this metadata. We have several strategies to make this less painful:
- if the query hasn't changed and we have metadata, don't do anything
- for mbql queries, we can just construct the metadata ourselves. We have field information, types, etc. So no need to hit a database and we just reason about the types
- for native queries, we have no such reasoning. We _must_ query the db. We attempt to make this nicer by running the query and abandoning it after the first row comes back. But we are still at the mercy of long-running queries even if we only take the first row.

#### The idea:
We proceed as before, starting an asynchronous fetching of the metadata. If, after some timeout period (randomly chosen as 1.5 seconds at first) we do not have metadata yet, allow the request to succeed and have the metadata populate the question whenever it is ready, provided that the query is still the same.

annotated Log output:
```
<--- send request
<--- wait 1.5 seconds
INFO api.card :: Metadata not available soon enough. Saving card 5,470 and asynchronously updating metadata
DEBUG middleware.log :: PUT /api/card/5470 200 1.5 s (18 DB calls) <----- request succeeds and unblocks UI
DEBUG middleware.log :: GET /api/alert/question/5470 200 1.0 ms
INFO api.card :: Metadata updated asynchronously for card 5,470     <----- later metadata is ready and saved in background
```

#### Hesitations

I'm a bit worried about this. There are lots of subtleties to metadata. And we have to untangle how the `Card` model has some safeguards already in place.

```clojure
(defn- populate-result-metadata
  "When inserting/updating a Card, populate the result metadata column if not already populated by inferring the
  metadata from the query."
  [{query :dataset_query, metadata :result_metadata, existing-card-id :id, :as card}]
  (cond
    ;; not updating the query => no-op
    (not query)
    (do
      (log/debug "Not inferring result metadata for Card: query was not updated")
      card)

    ;; passing in metadata => no-op
    metadata
    (do
      (log/debug "Not inferring result metadata for Card: metadata was passed in to insert!/update!")
      card)

    ;; this is an update, and dataset_query hasn't changed => no-op
    (and existing-card-id
         (= query (db/select-one-field :dataset_query Card :id existing-card-id)))
    (do
      (log/debugf "Not inferring result metadata for Card %s: query has not changed" existing-card-id)
      card)

    ;; query has changed (or new Card) and this is a native query => set metadata to nil
    ;;
    ;; we can't infer the metadata for a native query without running it, so it's better to have no metadata than
    ;; possibly incorrect metadata.
    (= (:type query) :native)
    (do
      (log/debug "Can't infer result metadata for Card: query is a native query. Setting result metadata to nil")
      (assoc card :result_metadata nil))

    ;; otherwise, attempt to infer the metadata. If the query can't be run for one reason or another, set metadata to
    ;; nil.
    :else
    (do
      (log/debug "Attempting to infer result metadata for Card")
      (let [inferred-metadata (not-empty (mw.session/with-current-user nil
                                           (classloader/require 'metabase.query-processor)
                                           (u/ignore-exceptions
                                             ((resolve 'metabase.query-processor/query->expected-cols) query))))]
        (assoc card :result_metadata inferred-metadata)))))
```

This is in the `pre-insert` and `pre-update` chain for `Card`s. This is a bit heavy handed (note for all native queries it just throws away the metadata _regardless_. My thinking is that this all needs to go away. But I need to research how all of this stuff fits together. It is quite sensitive and quite subtle.

So right now it blanks out the metadata when saving this for native queries. Perhaps we could thread through the previous metadata (if it exists) and we'd hit the case above.

##### Other considerations
this is inherently a bit racy. Especially if there are queries that can take longer than 15 minutes to complete. That's a lot of changes that can be done while we have a pending write of the metadata. Some occurrences:
- someone changes the query. Then the metadata that is pending is no longer even relevant. I check against this at the moment
- someone edits model metadata. Those edits will be tossed away at this time. At this point the metadata is actually empty so there's not much to edit. Not sure what the UI would do if you tried to edit in this case.
- there's an error when getting the metadata: connections can go down, etc. Before this change that would prevent saving and give us a chance to inform the user. Now it happens silently in a java process that no one will notice
- there's an error with the query. `select id, total, quantity, sleep(1) from orders limit 3 limit 3` is an invalid query but nothing really prevents that from saving right now. You can happily save it and just be frustrated later downstream from the actual issue.

dpsutton committed:
Lets use the main thread

dpsutton committed:
Strip out channel stuff and rename

dpsutton committed:
202 -> 200 response

When returning a channel we return a 202. A map is just a 200. Since we
no longer need to have the main stuff async (as opposed to the metadata
stuff) we can just return the map with a 200 instead of this long
running channel stuff and a 202.

dpsutton committed:
Last test

dpsutton committed:
renames, logging, ensure query is the same before saving metadata

dpsutton committed:
Sandbox test 202 -> 200

dpsutton committed:
Another 202 -> 200

adam-james-v commented:
> * someone edits model metadata. Those edits will be tossed away at this time. At this point the metadata is actually empty so there's not much to edit. Not sure what the UI would do if you tried to edit in this case.

Does editing 'nil' metadata in the UI just imply that a user is adding their own metadata? Could such data be handled with a merge of some kind once the metadata computation completes? `(merge inferred-metadata user-metadata)`? Of course, then a user's edits might make the inferred metadata completely obsolete and unnecessary...

> * there's an error when getting the metadata: connections can go down, etc. Before this change that would prevent saving and give us a chance to inform the user. Now it happens silently in a java process that no one will notice
> 
> * there's an error with the query. `select id, total, quantity, sleep(1) from orders limit 3 limit 3` is an invalid query but nothing really prevents that from saving right now. You can happily save it and just be frustrated later downstream from the actual issue.

These make me think of  a tuple of 'card health' `{:query-success? true :metadata-success? true}` with 'nil's or something while the query is pending. All of that could be used to report different information to the user? When both are true, things look exactly the same. For short queries, things would hopefully work alright. For something like a 15min query, there would be something to make that obvious to a user? It doesn't eliminate the problems, but can be useful to explain to users so that they know that cards are sort of 'conditionally saved'.

dpsutton committed:
Put timeout on async metadata saving

timeout of 15 minutes before we give up on the async metadata saving. It
is possible this cuts things off but hard to tell if work is still being
done at that point.

dpsutton commented:
So there's a failing e2e test.

The card endpoints used to return streaming responses so they _always_ returned a 202. Thus you had weirdness like the following:
<img width="1079" alt="image" src="https://user-images.githubusercontent.com/6377293/178369477-17ff2614-afa7-4fa4-8c7c-6502c5a40c25.png">

which is a 202 with payload:

```javascript
{"message":"Invalid Field Filter: Field 4 \"PRODUCTS\".\"CATEGORY\" belongs to Database 1 \"Sample Database\", but the query is against Database 2 \"copy sample dataset\"",
 "_status":400}
```

But after this change, the endpoint (sensibly IMO) returns a 400 with the message 
> Invalid Field Filter: Field 26 "PRODUCTS"."CATEGORY" belongs to Database 1 "Sample Database", but the query is against Database 990 "copy of sample dataset"


Functionally the same but more "proper". However, I need some help updating the FE to accept this new shape. I don't know if this is a big ask or not though.

To recreate:
- go to the admin > Databases screen and select the sample dataset. Copy it's path, create a new h2 database named "copy of sample dataset" and paste the path you copied earlier as the "Connection String"
- make a new native sql question based on the first sample dataset whose contents are `select count(*) from products where {{filter}}`. Make the variable `filter` a field filter on the category field from the products table
- on the top left, change the datasource from the sample dataset to the copy of the dataset
- Click save

We get the correct API behavior: a 400 and a nice error message, but the UI doesn't know to show it:
<img width="938" alt="image" src="https://user-images.githubusercontent.com/6377293/178370598-91ead3cd-081c-4669-bd62-607c57bedd8e.png">

alxnddr commented:
@dpsutton I'm wondering if you can make back-end not changing `content-type` to `text/plain` and return a JSON with error like 
```{"message":"Invalid Field Filter: Field 4 \"PRODUCTS\".\"CATEGORY\" belongs to Database 1 \"Sample Database\", but the query is against Database 2 \"copy sample dataset\""}```

dpsutton committed:
outdated comment

dpsutton committed:
Return json error message, not text/plain

this is a subtle one that I'm not happy about. Our error handling will
return text/plain if you throw an `(ex-info "something" {:status-code
400})`.

```shell
❯ http post localhost:3000/api/timeline-event/ name=some-name description=Bob timestamp=2022 timezone=America/Central time_matters:=false timeline_id:=1629  Cookie:$COOKIE
HTTP/1.1 404 Not Found
Content-Type: text/plain

Timeline with id 1,629 not found
```

But if you add extra information to the map to the ex-info, you get
json!

```clojure
(defmethod api-exception-response Throwable
  [^Throwable e]
  (let [{:keys [status-code], :as info} (ex-data e)
        other-info                      (dissoc info :status-code :schema :type)
        body                            (cond
                                          (and status-code (empty? other-info))
                                          ;; If status code was specified but other data wasn't, it's something like a
                                          ;; 404. Return message as the (plain-text) body.
                                          (.getMessage e)

                                          ;; if the response includes `:errors`, (e.g., it's something like a generic
                                          ;; parameter validation exception), just return the `other-info` from the
                                          ;; ex-data.
                                          (and status-code (:errors other-info))
                                          other-info

                                          ;; Otherwise return the full `Throwable->map` representation with Stacktrace
                                          ;; and ex-data
                                          :else
                                          (merge
                                           (Throwable->map e)
                                           {:message (.getMessage e)}
                                           other-info))]
    {:status  (or status-code 500)
     :headers (mw.security/security-headers)
     :body    body}))
```

So this fix is a _very_ subtle way to get to what we want, although it
does add a bunch of extra junk to our response.

```javascript
{
 ...
 "message": "Invalid Field Filter: Field 26 \"PRODUCTS\".\"CATEGORY\"
belongs to Database 1 \"Sample Database\", but the query is against
Database 990 \"copy of sample dataset\"",
 "data": {
   "status-code": 400,
   "query-database": 990,
   "field-filter-database": 1}
 ...
}
```

Reminder of what we want: the frontend is saving a card. We added a
field filter on a database and then changed the source database of the
query. So the backend needs to reject the field filter as being on the
wrong db. The FE expects a json response with a message and will then
show that message in the save/edit modal.

Why did this come up now: these endpoints for saving and editing a card
used to always return 202 streaming responses. This means they would
have to tuck any errors inside of an already open 202 response. Which is
why you get a message as a json response. But now that they are sync,
the api can just return a proper 400 with a reason. But we still want
that to be a json response for the FE.

alxnddr committed:
error layout fix

camsaul commented on src/metabase/api/card.clj:
```diff
@@ -245,31 +244,19 @@
+(def ^:private metadata-async-timeout
+  "Duration in milliseconds to wait for the metadata before abandoning the asynchronous metadata saving. Default is 15
+  minutes."
+  (* 1000 60 15))
```
We have `u/minutes->ms` which does the same thing but I think is a little easier to read and harder to accidentally forget a zero or whatever

camsaul commented on src/metabase/api/card.clj:
```diff
@@ -245,31 +244,19 @@
+saved later when it is ready."
+  1500)
+
+(def ^:private metadata-async-timeout
```
```suggestion
(def ^:private metadata-async-timeout-ms
```

Probably better to error on the side of overly explicit and make sure the fact that this is in milliseconds obvious just by looking at it. I know it's probably safe to assume it's in milliseconds since we didn't assume otherwise but I don't want someone who doesn't know that think this is in seconds and turn this into a 15,000 minute timeout

camsaul commented on src/metabase/api/card.clj:
```diff
@@ -280,24 +267,58 @@
+                             (log/info (trs "Not updating metadata asynchronously for card {0} because no metadata"
+                                            id))))
+                     (catch Exception e
+                       (log/warn e (trs "Error updating card metadata asynchronously"))))))))))
```
This is a mega function (66 lines) -- can any parts of it be split out into separate functions? It's a bit hard to read huge functions. But maybe the async metadata update stuff can be split out

camsaul commented on src/metabase/api/card.clj:
```diff
@@ -280,24 +267,58 @@
+                           [metadata port] (a/alts!! [result-metadata-chan timeoutc])
+                           id              (:id <>)
+                           current-query   (db/select-one-field :dataset_query Card :id id)]
+                       (cond (= port timeoutc)
```
I think you need to be closing `result-metadata-chan` here. Otherwise the query will keep running forever or until it times out in the QP

camsaul commented on src/metabase/api/card.clj:
```diff
@@ -280,24 +267,58 @@
+                          :collection [:moderation_reviews :moderator_details])
+                 (assoc :last-edit-info (last-edit/edit-information-for-user @api/*current-user*)))
+      (when timed-out?
+        (.submit clojure.lang.Agent/pooledExecutor
```
I'm not convinced `Agent/pooledExecutor` is the best executor to use. Its size is nCPUs + 2, so on a small EC2 instance it's 3 or 4. OTOH it's 34 on my home computer.

Neither one really matches up to the IRL effective limit of 15 open connections per data warehouse DB. 

My only concern is that on instances with a lot of users we could be waiting a while if a lot of people create Cards at the same time.

Since the QP already handles thread pool nonsense and connection pooling is handled by the drivers maybe just doing this with core async would be better. We already have `result-metadata-chan` anyway

dpsutton replied:
Good point. My main consideration here was to make sure we didn't run into shenanigans with toucan's connection closing. I can switch this around. Maybe I can finish waiting in a `go` block and then just submit somewhere (or even a future?) to actually save the the results to the db. That execution will need some thread somewhere but it its only concerned with saving (and not waiting on the channel) it should be fine. Good point.

camsaul replied:
👍 I/O should go on a separate thread so we don't tie up our precious limited core.async threads (I think the pool is fixed at like 4 or 8 threads or something) but `future` should be fine there.

I guess ideally maybe you want to do the `db/select` on a separate thread too for the same reason (you can use `a/thread` to run it on a different thread and get a channel to get the result when it comes back).

camsaul reviewed (approved):
Sorry this took so long to review.

> this is inherently a bit racy.

Do you mean race-conditiony? I like the idea of our code being vigorous and/or risqué but I don't think that's what you meant lol.

> there's an error with the query

This also applies for parameterized queries with required parameters -- we can't run them without those parameters and we don't have values for them when you save a Card.

With the writeback/actions stuff we're working on we're introducing the concept of `is_write` queries which you're not supposed to run at all ever outside of the new writeback QP -- but it will never have result metadata. You don't need to worry about this here but in the future we'll probably have more cases where people are saving queries without running them, that can't be ran normally either

### Anyways

Anyways overall this looks like a big improvement over the current state of things. I think this also resolves #14385, right?

dpsutton committed:
Several Cleanups

- make sure numbers have units at the end (-ms)
- use (u/minutes->ms 15) rather than a more opaque (* 15 60 1000)
- move the scheduled metadata saving into its own function

This last bit is actually a bit more than that. I was previously
throwing everything in a thread submitted to the pooled executor. I'm
now using a `future` but with a caveat: All of the waiting for the
timeout, checking if we got metadata is done in a simple `a/go` block
now, and the thread does just the last bit of IO. We have to select the
card as it is now to ensure the query is the same and then save.

Refresher on why we have to check if the query is the same. We have up
to 15 minutes to wait for the query metadata to come back. Completely
possible for them to have edited the query and then our metadata is
useless.
